Fix: other courses

Illia-Sakharau opened the pull request:
## What type of PR is this? (select all that apply)

- [x] 🐛 Bug Fix

## Description
- [x] Return nearest course to Upcoming courses
- [x] Refactor component `Courses`
- [x] Update start date for courses
- [x] Change label text on course page (in `CourseMain` component)

## Related Tickets & Documents
- Related Issue #44 
- Closes #44 

- Related Issue #48 
- Closes #48 

## Screenshots, Recordings
<img width="1333" alt="Screenshot 2024-04-16 at 16 54 55" src="https://github.com/rolling-scopes/site/assets/124388500/ccbdbdcf-78f3-43ea-a3ef-3f38ab4f607f">


## Added/updated tests?

- [x] 👌 Yes
- [ ] 🙅‍♂️ No, because they aren't needed
- [ ] 🙋‍♂️ No, because I need help

andron13 reviewed (changes requested):
1. The component seems quite large and does quite a few tasks. You might want to consider breaking it down into smaller components, for instance, CourseList and CourseItem.
2. I have to say, I found **courses as Course[]** to be less than ideal.
3. What's going to happen if the data array turns out to be empty?
4. We seem to be parsing the date twice over here. Is there a way we could avoid this duplicate use of `Date.parse`?

Illia-Sakharau commented:
1. Removed all unnecessary items from the component.
2. Replaced with a data type check.
3. Added the processing of this case (empty data array).
4. This is not a duplicate of `Date.parse`, I will convert the two values being compared.

andron13 commented on src/app/hooks/use-nearest-course/utils/choose-nearest-course.ts:
```diff
@@ -0,0 +1,22 @@
+  const dateNow = Date.now();
+  let course = !nextCourse ? prevCourse : nextCourse;
+  if (nextCourse && prevCourse) {
+    const bufferPeriodMs = bufferPeriod * 24 * 60 * 60 * 1000;
```
А зачем переменные создавать в проверке условия?

Illia-Sakharau replied:
Изменил

andron13 commented on src/app/hooks/use-nearest-course/utils/define-prev-next.ts:
```diff
@@ -0,0 +1,31 @@
+  let nextCourse: Course | undefined;
+
+  coursesData.forEach((obj: object) => {
+    if(isCourse(obj)) {
+      const startDate = Date.parse(obj.startDate);
+      const isPast = startDate <= dateNow;
+      if (
+        (!prevCourse && isPast) 
+        || (isPast && Date.parse((prevCourse as Course).startDate) < startDate)
+      ) {
+        prevCourse = obj;
+      }
+      if (
+        !nextCourse && !isPast
+        || (!isPast && Date.parse((nextCourse as Course).startDate) > startDate)
+      ) {
+        nextCourse = obj;
+      }
+    }
```
я не смог это прочитать.

Illia-Sakharau replied:
Рефакторинг запланирован в таске [#85](https://github.com/rolling-scopes/site/issues/85)

andron13 commented on src/features/courses/courses/courses.tsx:
```diff
@@ -1,55 +1,43 @@
                   key={course.id}
                   {...course}
-                  startDate={ dateNow < Date.parse(course.startDate) ? course.startDate : '(Upcoming)' }
+                  startDate={ nearestCourseStartDate <= Date.parse(course.startDate) ? course.startDate : '(TBD)' }
```
я не приму в рендере расчёты.

Illia-Sakharau replied:
убрал расчеты из рендера
